activerecord/test/cases/test_case.rb

```diff
@@ -30,6 +30,25 @@ def teardown
       SQLCounter.clear_log
     end
 
+    def before_setup
```

Comment:
The change we did in this PR revealed a bunch of leaks in the ActiveRecord test suite. One of the main issue
was tests doing something like

```ruby
test 'foo' do
  ActiveRecord::Base.connects_to(writing: :arunit2, reading: :other_db)
end
```

The problem with this is that after that test runs, `ActiveRecord::Base` connection_handler will hold the connection on the activerecord_unittest2 db (and not the regular activerecord_unittest).

I started fixing tests independently but ended up in a huge rabbit hole so I opted for a "global" state reset instead of doing a `@previous_connection_handlers = ..., ensure ...` in a **lot** of places